quinn/src/endpoint.rs

```diff
@@ -253,39 +253,43 @@ where
 {
     fn drive_recv(&mut self, cx: &mut Context, now: Instant) -> Result<bool, io::Error> {
         let mut recvd = 0;
+        let mut metas = [RecvMeta::default(); BATCH_SIZE];
+        let mut iovs = [IoSliceMut::new(&mut [])];
+        for (iov, buf) in iovs
+            .iter_mut()
+            .zip(self.recv_buf.chunks_mut(self.recv_buf.len() / BATCH_SIZE))
+        {
+            *iov = IoSliceMut::new(buf);
+        }
         loop {
-            let mut meta = [RecvMeta::default()];
-            match self
-                .socket
-                .poll_recv(cx, &mut [IoSliceMut::new(&mut self.recv_buf)], &mut meta)
-            {
+            let result = self.socket.poll_recv(cx, &mut iovs, &mut metas);
+            match result {
                 Poll::Ready(Ok(msgs)) => {
-                    debug_assert_eq!(msgs, 1);
-                    match self.inner.handle(
-                        now,
-                        meta[0].addr,
-                        meta[0].ecn,
-                        (&self.recv_buf[0..meta[0].len]).into(),
-                    ) {
-                        Some((handle, DatagramEvent::NewConnection(conn))) => {
-                            let conn = self.connections.insert(handle, conn);
-                            if self.incoming_live {
-                                self.incoming.push_back(conn);
-                                if let Some(task) = self.incoming_reader.take() {
-                                    task.wake();
+                    for (meta, buf) in metas.iter().zip(iovs.iter()).take(msgs) {
+                        match self
+                            .inner
+                            .handle(now, meta.addr, meta.ecn, buf[0..meta.len].into())
+                        {
+                            Some((handle, DatagramEvent::NewConnection(conn))) => {
+                                let conn = self.connections.insert(handle, conn);
```

Comment:
So here we're at `loop` / `match` / `for` / `match` (and then we still get two levels of `if`). Is there a clean way to make this less deeply nested, perhaps by just extracting some function from the inside, even it has just a single use?

Reply:
This is complicated by borrowck. `iovs` mutably borrows from `self`, so we can't easily extract a method on `EndpointInner`.